src/pybennu/pybennu/providers/power/solvers/siren.py

```diff
@@ -0,0 +1,493 @@
+            server_endpoint: IP address for the provider server to listen on
+            publish_endpoint: IP address for the provider to publish values to
+            settings: (PybennuSettings) The settings found in the siren_config.yaml 
+                      file generated from the scenario yaml
+        """
+        super().__init__(server_endpoint, publish_endpoint)
+
+        self.conf: PybennuSettings = settings
+        self.setup_logger()
+
+        siren_config = self.conf.siren.siren_json
+        passed_devices = None
+        self.setup_siren_config(siren_config, passed_devices) # Populates self.device_configs (dict)
+
+        self.devices = {}
+        self.to_endpoint_states = {} # Holds hardware values before written to internal_state
+        self.to_hardware_threads = []
+        self.from_hardware_threads = []
+
+        for device_name, dev_config in self.device_configs.items():
+
+            self.devices[device_name] =  self.create_device(dev_config)
+
+            self.to_endpoint_states[device_name] = dict.fromkeys(dev_config.to_endpoint.keys())
+
+            # Writes values to hardware from internal state.
+            # Replaces _subscription_handler functionality
+            if dev_config.to_device.keys():
+                thread = threading.Thread(target=self.write_to_hardware, args=(device_name,)) 
+                thread.daemon = True
+                self.to_hardware_threads.append(thread)
+
+            # Writes value to internal state from hardware
+            # Replaces _to_provider functionality
+            if dev_config.to_endpoint.keys():
+                thread = threading.Thread(target=self.update_from_hardware, args=(device_name,))
+                thread.daemon = True
+                self.from_hardware_threads.append(thread)
+
+        self.run_thread = True
+
+        
+
+        self.IS_lock = threading.Lock()
+        self.set_internal_state()
+
+        
+    def set_internal_state(self, ):
+        """
+        Create and Initialize the internal state dictionary of the Provider.
+
+        This self.internal_state variale should be only be accessed with the self.IS_lock
+        """
+        self.internal_state = {}
+        self.ac_values = {}
+        self.rms_tags = set()
+        for tag in self.conf.siren.tags:
+            self.internal_state[tag.name] = str(tag.initial_value)
+            if tag.publish_rms:
+                self.rms_tags.add(tag.name)
+                self.ac_values[tag.name] = CircularBuffer(self.conf.siren.rms_buffer_size)
+
+
+    def setup_siren_config(self, config, devices):
+        """
+        Create device config dictionaries for each device in the siren.json
+
+        Args:
+            config: (str) Path to the siren.json config file
+            devices: Specific devices to create 
+        """
+        with open(config, "r") as f:
+            self.siren_config = json.load(f)
+
+        self.device_configs = {}
+        for key in self.siren_config:
```

Comment:
Use `.items()` here, e.g. for `for key, conf in self.siren_config.items()`, then replace `self.siren_config[key]` with `conf`.

Also, I feel like there's a better way to do this instead of hardcoding key to attribute conversions in `if` statements. Maybe with [Box](https://github.com/cdgriffith/Box/wiki/Quick-Start) (which is already used in phenix-apps). That's a issue for another day.